Cleanup minor environment variable issues

tobil4sk opened the pull request:
This was an oversight in #10402. Running `Sys.putEnv("var", null)` would remove the variable from the internal map used to keep track of environment variables, so it would appear that the variable was gone (`Sys.getEnv()` calls would also suggest that). However, the variable would not actually be removed, and would still exist in child processes for example.This fixes that issue on python and adds a `python.lib.Os.unsetenv()` extern which is needed to properly remove environment variables.

Also I added a test to ensure everywhere that removed variables are gone for good (by checking if they exist in child processes).

Fixes #10469 - Changes made via the Haxe API are visible when using the native Python/C# apis and vice versa.
Fixes #10471 - Sys.environment() returns only valid variables on php and python.

tobil4sk committed:
[python] Ensure putEnv properly removes variables

Requires os.unsetenv extern

tobil4sk committed:
[tests] Add extra testing for removing env variables

Previously the tests did not ensure that the variables would
be updated in child processes

tobil4sk commented:
hm... python for windows doesn't have this function prior to 3.9.

RealyUniqueName commented on tests/sys/src/TestSys.hx:
```diff
@@ -32,6 +32,10 @@ class TestSys extends TestCommandBase {
 		#end
 	}
 
+	function existsInSubProcess(variable:String, value:String) {
```
`value` is not used

tobil4sk replied:
Thanks

RealyUniqueName commented:
> hm... python for windows doesn't have this function prior to 3.9.

In that case we can at least remove the variable from internal maps (Haxe's `Sys.environ` and Python's `os.environ`)

tobil4sk commented:
The python documentation says that the `os.unsetenv()` function was available prior to 3.9 in "most flavours of unix" which isn't very specific.

`Sys.putEnv()` already updates haxe's `Sys.environ()`, however, when setting a variable it doesn't modify python's `os.environ` (it calls `os.putenv()`, which is specifically said to not update `os.environ`) so doing that for removing the variable would be inconsistent, unless that should change as well. 

The situation on python is a bit messy in itself:
https://docs.python.org/3.10/library/os.html#os.environ

tobil4sk committed:
[python] Fix Sys.putEnv for python 3.8 and below

Add a try catch block so that this function doesn't throw when targeting
python versions prior to 3.9, as before that `os.unsetenv` is not
guaranteed to exist

tobil4sk committed:
Document putEnv(var, null) limitation on python

tobil4sk committed:
[tests] Update sys tests accordingly

tobil4sk commented:
The `os.unsetenv` call is now wrapped in a try catch block so that it runs wherever it is supported.

lublak commented:
Just for a question: can you not just call del os.environ.pop('env') I think python auto call unsetenv if it's available.

tobil4sk commented:
That's another way to do it, however, right now haxe's `Sys.putEnv()` hasn't been messing at all with python's `os.environ`. So it would require specifying that behaviour for all usage of the function.

lublak commented:
I think its the best solution. Most python Devs use osml.environ directly.
I also readed that unsetenv has an bug on some osx platforms, but del os.environ works fine.

tobil4sk commented:
Alright, having looked into it I now do agree that this is the better option, as it removes the need for the try catch block because whether the unsetenv function exists is handled on the python side. Also, looking at the docs for `os.putenv()` for python 3.8, it also has the note that it is "available on most flavours of unix", so we don't even know if that exists necessarily. By using `os.environ` instead of the direct functions it might be a bit safer in terms of compatiblity. There's just some things left to consider about how far python's and haxe api's need to be kept compatible.

tobil4sk committed:
[python] Remove internal environment variable map

Fixes #10469

Sys.putEnv() now modifies python's `os.environ` dictionary rather than
using `os.putenv` or `os.unsetenv` and modifying an internal map. This
is a better choice as modifying the dictionary calls these functions
automatically if they exist.
This also means that changes made via the Haxe api will be visible also
via python's `os.environ`.

As a result, Sys no longer needs to store an internal map of environment
variables. As of #10460, a copy is made of the map
anyway everytime `Sys.environment()` is called. This means we might as
well directly loop through python's `os.environ` which takes into
account modifications made via the python
api, rather than looping through an internal copy which does not, and
takes the same amout of time.

tobil4sk committed:
[cs] Remove internal copy of environment variables

Fixes #10469

Now, modifications to the environment directly via the
native C# api will be visible in 'Sys.environment()`

tobil4sk commented:
The issue might even be resolved by using `os.environ`, at least from my testing I was able to get around the issue of `os.unsetenv` not existing completely just by removing it directly from `os.environ`. It seems like python uses `os.environ` when deciding which variables should be passed on to child processes.

tobil4sk commented:
Also using `os.environ` directly should address the python side of #10469, and I've also cleaned up C# in a similar way to address the same issue, because with #10460 it is a really trivial change.

lublak commented:
Just some extra information:
environ.remove throws an exception if the environment not exists. (KeyError)
Just use pop or check if the key exists

tobil4sk committed:
[python] Allow removing non-existent variable

tobil4sk committed:
[tests] Test removal of non-existent variable

tobil4sk commented:
@lublak Thanks, that's a good catch.

tobil4sk commented:
Also fixes #10471.

tobil4sk committed:
[python] Remove warning about putEnv(name, null)

This should not be an issue now that Sys.putEnv modifies `os.environ`
directly

tobil4sk committed:
[tests] Only test uppercase environment variables

For compatibility with python

tobil4sk committed:
[tests] Test getEnv case-insensitivity on windows

tobil4sk committed:
[php] Remove incorrect environment variables

Fixes #10471

On windows, ensure only correct and up to date environment variables are
present in `Sys.environment()`

tobil4sk committed:
[sys] Document Sys.environment() capitalization
